// @allurereport/web-commons/dist/charts/colors.js
export const statusColors = {
    failed: "var(--bg-support-capella)",
    broken: "var(--bg-support-atlas)",
    passed: "var(--bg-support-castor)",
    skipped: "var(--bg-support-rau)",
    unknown: "var(--bg-support-skat)",
};
export const severityColors = {
    blocker: "var(--bg-support-capella)",
    critical: "var(--bg-support-atlas)",
    normal: "var(--bg-support-castor)",
    minor: "var(--bg-support-rau)",
    trivial: "var(--bg-support-skat)",
};
export const statusChangeColors = {
    newFailed: "var(--bg-support-capella)",
    newBroken: "var(--bg-support-atlas)",
    newPassed: "var(--bg-support-castor)",
    newSkipped: "var(--bg-support-rau)",
    newUnknown: "var(--bg-support-skat)",
    removedFailed: "color-mix(in srgb, var(--bg-support-capella) 80%, black)",
    removedBroken: "color-mix(in srgb, var(--bg-support-atlas) 80%, black)",
    removedPassed: "color-mix(in srgb, var(--bg-support-castor) 80%, black)",
    removedSkipped: "color-mix(in srgb, var(--bg-support-rau) 80%, black)",
    removedUnknown: "color-mix(in srgb, var(--bg-support-skat) 80%, black)",
};
export const generateLayerColors = (layers, baseColor = "var(--bg-support-aldebaran)") => {
    const colors = {};
    if (layers.length === 0) {
        return colors;
    }
    if (layers.length === 1) {
        colors[layers[0]] = baseColor;
        return colors;
    }
    layers.forEach((layer, index) => {
        const ratio = index / (layers.length - 1);
        if (ratio === 0) {
            colors[layer] = `color-mix(in srgb, ${baseColor} 60%, white)`;
        }
        else if (ratio === 1) {
            colors[layer] = `color-mix(in srgb, ${baseColor} 80%, black)`;
        }
        else {
            const lightRatio = 60 + 20 * (1 - ratio);
            const darkRatio = 20 * ratio;
            colors[layer] =
                `color-mix(in srgb, color-mix(in srgb, ${baseColor} ${lightRatio}%, white) ${100 - darkRatio}%, black)`;
        }
    });
    return colors;
};
export const resolveCSSVarColor = (value, el = document.documentElement) => {
    if (value.startsWith("var(")) {
        const match = value.match(/var\((--[^),\s]+)/);
        if (match) {
            const cssVarName = match[1];
            const resolved = getComputedStyle(el).getPropertyValue(cssVarName).trim();
            return resolved || value;
        }
    }
    return value;
};
