// @allurereport/charts-api/dist/types.js
export var ChartType;
(function (ChartType) {
    ChartType["Trend"] = "trend";
    ChartType["Pie"] = "pie";
    ChartType["TreeMap"] = "treemap";
    ChartType["HeatMap"] = "heatmap";
    ChartType["Bar"] = "bar";
    ChartType["Funnel"] = "funnel";
    ChartType["ComingSoon"] = "coming-soon";
})(ChartType || (ChartType = {}));
export var ChartDataType;
(function (ChartDataType) {
    ChartDataType["Status"] = "status";
    ChartDataType["Severity"] = "severity";
})(ChartDataType || (ChartDataType = {}));
export var BarChartType;
(function (BarChartType) {
    BarChartType["StatusBySeverity"] = "statusBySeverity";
    BarChartType["StatusTrend"] = "statusTrend";
    BarChartType["StatusChangeTrend"] = "statusChangeTrend";
    BarChartType["DurationsByLayer"] = "durationsByLayer";
    BarChartType["FbsuAgePyramid"] = "fbsuAgePyramid";
    BarChartType["StabilityRateDistribution"] = "stabilityRateDistribution";
})(BarChartType || (BarChartType = {}));
export var TreeMapChartType;
(function (TreeMapChartType) {
    TreeMapChartType["SuccessRateDistribution"] = "successRateDistribution";
    TreeMapChartType["CoverageDiff"] = "coverageDiff";
})(TreeMapChartType || (TreeMapChartType = {}));
export var ChartMode;
(function (ChartMode) {
    ChartMode["Raw"] = "raw";
    ChartMode["Percent"] = "percent";
    ChartMode["Diverging"] = "diverging";
})(ChartMode || (ChartMode = {}));
export var BarGroupMode;
(function (BarGroupMode) {
    BarGroupMode["Grouped"] = "grouped";
    BarGroupMode["Stacked"] = "stacked";
})(BarGroupMode || (BarGroupMode = {}));
