// @allurereport/web-commons/dist/charts/bar.js
import { BarChartType, ChartMode, DEFAULT_CHART_HISTORY_LIMIT } from "@allurereport/charts-api";
import { statusBySeverityBarDataAccessor } from "./accessors/statusBySeverityBarAccessor.js";
import { statusChangeTrendBarAccessor } from "./accessors/statusChangeTrendBarAccessor.js";
import { statusTrendBarAccessor } from "./accessors/statusTrendBarAccessor.js";
import { generateBarChartDurationsByLayer } from "./bar/generateBarChartDurationsByLayer.js";
import { generateBarChartFbsuAgePyramid } from "./bar/generateBarChartFbsuAgePyramid.js";
import { generateStabilityRateDistribution } from "./bar/generateStabilityRateDistribution.js";
import { limitHistoryDataPoints } from "./chart-utils.js";
export const generateBarChartGeneric = (options, storeData, dataAccessor) => {
    const { type, dataType, title, limit = DEFAULT_CHART_HISTORY_LIMIT, mode = ChartMode.Raw } = options;
    const { historyDataPoints } = storeData;
    const limitedHistoryPoints = limitHistoryDataPoints(historyDataPoints, limit);
    const isFullHistory = limitedHistoryPoints.length === historyDataPoints.length;
    const items = dataAccessor.getItems(storeData, limitedHistoryPoints, isFullHistory);
    let processedData = items;
    if (mode === ChartMode.Percent) {
        processedData = items.map((group) => {
            const { groupId, ...values } = group;
            const total = Object.values(values).reduce((sum, value) => sum + value, 0);
            const nextValues = Object.keys(values).reduce((acc, valueKey) => {
                acc[valueKey] = values[valueKey] / total;
                return acc;
            }, {});
            return {
                groupId,
                ...nextValues,
            };
        });
    }
    return {
        type,
        dataType,
        mode,
        title,
        data: processedData,
        keys: dataAccessor.getGroupKeys(),
        groupMode: dataAccessor.getGroupMode(),
        indexBy: "groupId",
    };
};
export const generateBarChart = (options, storeData) => {
    const newOptions = { limit: DEFAULT_CHART_HISTORY_LIMIT, ...options };
    const { dataType } = newOptions;
    switch (dataType) {
        case BarChartType.StatusBySeverity:
            return generateBarChartGeneric(newOptions, storeData, statusBySeverityBarDataAccessor);
        case BarChartType.StatusTrend:
            return generateBarChartGeneric(newOptions, storeData, statusTrendBarAccessor);
        case BarChartType.StatusChangeTrend:
            return generateBarChartGeneric(newOptions, storeData, statusChangeTrendBarAccessor);
        case BarChartType.DurationsByLayer:
            return generateBarChartDurationsByLayer(newOptions, storeData);
        case BarChartType.FbsuAgePyramid:
            return generateBarChartFbsuAgePyramid(newOptions, storeData);
        case BarChartType.StabilityRateDistribution:
            return generateStabilityRateDistribution(newOptions, storeData);
    }
};
