// @allurereport/web-commons/dist/charts/bar/generateBarChartFbsuAgePyramid.js
import { BarChartType, BarGroupMode, ChartMode, ChartType, DEFAULT_CHART_HISTORY_LIMIT, } from "@allurereport/charts-api";
import { limitHistoryDataPoints } from "../chart-utils.js";
const generateDataPoint = (testResults, name) => {
    let failed = 0;
    let broken = 0;
    let skipped = 0;
    let unknown = 0;
    for (const testResult of testResults) {
        switch (testResult.status) {
            case "failed":
                failed++;
                break;
            case "broken":
                broken++;
                break;
            case "skipped":
                skipped++;
                break;
            case "unknown":
                unknown++;
                break;
        }
    }
    const totalTests = failed + broken + skipped + unknown;
    const failedPercent = totalTests > 0 ? failed / totalTests : 0;
    const brokenPercent = totalTests > 0 ? broken / totalTests : 0;
    const skippedPercent = totalTests > 0 ? skipped / totalTests : 0;
    const unknownPercent = totalTests > 0 ? unknown / totalTests : 0;
    const dataPoint = {
        groupId: name,
    };
    dataPoint.failed = Math.round(-failedPercent * 100);
    dataPoint.broken = Math.round(-brokenPercent * 100);
    dataPoint.skipped = Math.round(skippedPercent * 100);
    dataPoint.unknown = Math.round(unknownPercent * 100);
    return dataPoint;
};
export const generateBarChartFbsuAgePyramid = (options, storeData) => {
    const { title, limit = DEFAULT_CHART_HISTORY_LIMIT } = options;
    const { historyDataPoints, testResults } = storeData;
    const limitedHistoryPoints = limitHistoryDataPoints(historyDataPoints, limit);
    const chartData = [
        generateDataPoint(testResults, "current"),
        ...limitedHistoryPoints.map((historyPoint, index) => generateDataPoint(Object.values(historyPoint.testResults), `Point ${index + 1}`)),
    ];
    return {
        data: chartData,
        type: ChartType.Bar,
        dataType: BarChartType.FbsuAgePyramid,
        mode: ChartMode.Diverging,
        title,
        keys: ["failed", "broken", "skipped", "unknown"],
        groupMode: BarGroupMode.Stacked,
        indexBy: "groupId",
        xAxisConfig: {
            legend: "Percentage of Tests",
        },
        yAxisConfig: {
            legend: "",
            format: "preserve",
            enabled: false,
            domain: [-100, 100],
        },
        layout: "horizontal",
    };
};
