// @allurereport/web-commons/dist/charts/accessors/statusBySeverityBarAccessor.js
import { BarGroupMode } from "@allurereport/charts-api";
import { severityLabelName, severityLevels, statusesList } from "@allurereport/core-api";
const processTestResults = (testResults) => {
    const resultMap = {
        blocker: undefined,
        critical: undefined,
        normal: undefined,
        minor: undefined,
        trivial: undefined,
    };
    severityLevels.forEach((severity) => {
        resultMap[severity] = statusesList.reduce((acc, status) => ({ ...acc, [status]: 0 }), {});
    });
    testResults.forEach((test) => {
        const severityLabel = test.labels?.find((label) => label.name === severityLabelName);
        const severity = severityLabel?.value?.toLowerCase();
        if (severity && resultMap[severity]) {
            resultMap[severity][test.status] = (resultMap[severity][test.status] ?? 0) + 1;
        }
    });
    return Object.entries(resultMap).reduce((acc, [severity, values]) => {
        if (values) {
            acc.push({ groupId: severity, ...values });
        }
        return acc;
    }, []);
};
export const statusBySeverityBarDataAccessor = {
    getItems: ({ testResults }) => {
        return processTestResults(testResults);
    },
    getGroupKeys: () => statusesList,
    getGroupMode: () => BarGroupMode.Grouped,
};
